src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcTaskHost.cs

```diff
@@ -374,51 +374,51 @@ internal static void ClearCachedTaskHostPaths()
         /// </summary>
         internal static string GetTaskHostNameFromHostContext(HandshakeOptions hostContext)
         {
-            ErrorUtilities.VerifyThrowInternalErrorUnreachable((hostContext & HandshakeOptions.TaskHost) == HandshakeOptions.TaskHost);
-            if ((hostContext & HandshakeOptions.CLR2) == HandshakeOptions.CLR2)
+            ErrorUtilities.VerifyThrowInternalErrorUnreachable(IsHandshakeOptionEnabled(HandshakeOptions.TaskHost));
+
+            if (IsHandshakeOptionEnabled(HandshakeOptions.CLR2))
             {
                 return TaskHostNameForClr2TaskHost;
             }
             else
             {
                 if (s_msbuildName == null)
                 {
-                    s_msbuildName = Environment.GetEnvironmentVariable("MSBUILD_EXE_NAME");
-
-                    if (s_msbuildName == null)
-                    {
-                        s_msbuildName = (hostContext & HandshakeOptions.NET) == HandshakeOptions.NET
-                            ? "MSBuild.dll"
-                            : "MSBuild.exe";
-                    }
+                    // for NET the executable is resolved from DOTNET_EXPERIMENTAL_HOST_PATH
+                    s_msbuildName = IsHandshakeOptionEnabled(HandshakeOptions.NET)
+                        ? string.Empty
+                        : Environment.GetEnvironmentVariable("MSBUILD_EXE_NAME") ?? Constants.MSBuildExecutableName;
```

Comment:
The comment says DOTNET_EXPERIMENTAL_HOST_PATH but the code is using MSBUILD_EXE_NAME. Not quite following that bit.

Reply:
DOTNET_EXPERIMENTAL_HOST_PATH is set as a project property that we will read later in [TaskBuilder.cs](https://github.com/dotnet/msbuild/pull/11543/files/98324e95e11464a0d2ea6cf7bf16363122c8d58d#diff-670154276685f1ec88f8e9537063d667c06005cace1ce2aa75e5bb61cc091ed6)

previously for HandshakeOptions.NET it was hardcoded here as 'MSBuild.dll'

Reply:
I'm still missing though what DOTNET_EXPERIMENTAL_HOST_PATH has to do with this line of code.

Reply:
I see your concern, I will update the comment for it.